### PR TITLE
Add unit tests for the Android app.

### PR DESCRIPTION
**Pull request by Isira-Seneviratne**
Fixes #

**Description of the Change**
Add unit tests for the parser classes that are used in the Android app, and make the changes needed for the unit tests. This allows any breaking changes made to the parser classes to be identified during the CI process.

**Release Notes**
N/A

### CODE REVIEW

**Comment by Isira-Seneviratne** [comment]
> Please move all tests to separate folder to be not mixed with the main code

They're already in a separate folder (src/test/java).

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java [hunk 1]
```diff
@@ -541,16 +544,16 @@ else if (Logging.DEBUG)
      *
      * @return list of attachable projects
      */
-    public ArrayList<ProjectInfo> getAttachableProjects(String boincPlatformName, String boincAltPlatformName) {
+    public List<ProjectInfo> getAttachableProjects(String boincPlatformName, String boincAltPlatformName) {
         if (Logging.DEBUG)
             Log.d(Logging.TAG, "getAttachableProjects for platform: " + boincPlatformName + " or " + boincAltPlatformName);
 
         ArrayList<ProjectInfo> allProjectsList = getAllProjectsList(); // all_proejcts_list.xml
```
[the remark this answers is not in the document]
Comment (Isira-Seneviratne) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientInterfaceImplementation.java]: I missed that one, thanks!

I'll convert it to List as well.

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientNotification.java [hunk 2]
```diff
@@ -66,24 +67,23 @@ public ClientNotification(Context ctx) {
...
+            List<Result> activeTasks = updatedStatus.getExecutingTasks();
             if(activeTasks.size() != mOldActiveTasks.size()) {
                 activeTasksChanged = true;
             }
             else {
                 for(int x = 0; x < activeTasks.size(); x++) {
                     if(!activeTasks.get(x).name.equals(mOldActiveTasks.get(x).name)) {
                         activeTasksChanged = true;
-                        Log.d("blub", "bla " + activeTasks.get(x).name + " vs. " + mOldActiveTasks.get(x).name); //TODO
+                        Log.d("blub", "bla " + activeTasks.get(x).name + " vs. " + mOldActiveTasks.get(x).name);
```
[the remark this answers is not in the document]
Comment (Isira-Seneviratne) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/client/ClientNotification.java]: Sure.

**Comment by AenBleidd** [comment]
> > Please move all tests to separate folder to be not mixed with the main code
> 
> They're already in a separate folder (src/test/java).

My mistake. You're right

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/PlatformInfo.java [hunk 3]
```diff
@@ -64,4 +64,12 @@ public PlatformInfo createFromParcel(Parcel in) {
         }
     };
 
+    @SuppressWarnings("java:S115")
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/PlatformInfo.java]: What this warning is about? I don't like to suppress warning. Is it possible to have no warning here?

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfig.java [hunk 4]
```diff
@@ -121,7 +120,27 @@ public ProjectConfig createFromParcel(Parcel in) {
         }
 
         public ProjectConfig[] newArray(int size) {
-            return null;
+            return new ProjectConfig[size];
         }
     };
+
+    @SuppressWarnings("java:S115")
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectConfig.java]: The same for this warning

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectInfo.java [hunk 5]
```diff
@@ -62,36 +82,34 @@ public void writeToParcel(Parcel dest, int arg1) {
...
+                public ProjectInfo[] newArray(int size) {
+                    return new ProjectInfo[size];
+                }
+            };
 
-    public static final Parcelable.Creator<ProjectInfo> CREATOR = new Parcelable.Creator<ProjectInfo>() {
-        public ProjectInfo createFromParcel(Parcel in) {
-            return new ProjectInfo(in);
-        }
+    @SuppressWarnings("java:S115")
```
Comment (AenBleidd) [inline android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/ProjectInfo.java]: The same for this warning

File: android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/PlatformInfo.java (hunk 3 above)
Comment (Isira-Seneviratne) [reply android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/PlatformInfo.java]: It's a warning provided by SonarLint. It relates to the naming of constant variables.

**Comment by AenBleidd** [comment]
As far as I see, currently unit tests are not executed on CI. Could you please fix this?

**Comment by ChristianBeer** [comment]
The build on Travis CI seems to fail with:
```
aidl E 03-23 10:58:01  9756  9756 aidl.cpp:581] refusing to generate code from aidl file defining parcelable
```
although it is not flagged as failing.

When are the android unit tests supposed to run? Before the actual build or are there extra commands needed?

@AenBleidd it seems we both had he same thought at the same time :). I never implemented a deploy stage for android (to check build products) and most likely won't have time before September to look into that if at all. I'll leave this in your capable hands.

**Comment by AenBleidd** [comment]
@ChristianBeer, thank you for pointing me to failed aidls. I didn't mention this.
Yes, I'll take care about all this together with @Isira-Seneviratne

**Comment by Isira-Seneviratne** [comment]
> As far as I see, currently unit tests are not executed on CI. Could you please fix this?

Output from a test is only displayed when it fails in the default configuration.

I'll add a configuration to display all tests, whether they pass or fail.

**Comment by AenBleidd** [comment]
Close and reopen PR to retrigger CI

**Pull request merged**